README.md
# pdf/stamper - PDF Templates
  http://github.com/jaywhy/pdf-stamper/
  by Jason Yates

## Description

Fill out PDF forms (templates) using iText's PdfStamper.

Features include:

* Text fields
* Checkboxes
* Circles, ellipses, rectangles
* PDF-417 Barcodes

You have to use JRuby or RJB. You need Adobe LiveCycle Designer
or Acrobat Professional to create the templates.

## Example
    pdf = PDF::Stamper.new("my_template.pdf")
  
    pdf.text :first_name, "Jason"
    pdf.text :last_name, "Yates"
  
    pdf.image :photo, "photo.jpg"
    pdf.checkbox :hungry
    pdf.ellipse(140, 380, 50, 13)
    pdf.rectangle(140, 380, 50, 13)
    pdf.circle(140, 380)
    
    # PDF-417 barcodes
    pdf.barcode("PDF417", "2d_barcode", "Barcode data...", AspectRatio: 0.5)
    
    pdf.save_as "my_output.pdf"

Here is how you use it in Rails.

    # In your controller.
    def send 
      pdf = PDF::Stamper.new("sample.pdf") 
      pdf.text :first_name, "Jason"
      pdf.text :last_name, "Yates" 
      send_data(pdf.to_s, :filename => "output.pdf", :type => "application/pdf",:disposition => "inline")
    end   

## Installation

  gem install pdf-stamper

## Code

  git clone http://github.com/jaywhy/pdf-stamper/

## Contributors

Paul Schreiber
Joe Steele
Ronny Hanssen
Marshall Anschutz

## License

(The MIT License)

Copyright (c) 2007-2013 Jason Yates

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
'Software'), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


History.txt
== 0.6.0 2013-10-28
* Adding PDF 417 Barcodes
* Cleaning up some other changes since last official release of 0.3.3.
* Checkboxes now work in JRuby (from https://github.com/paulschreiber/pdf-stamper/commit/e5098c670aa4b2b387c66411fa7b3f63c6511224)


== 0.3.3 2010-09-24
* iText downgraded to 4.2.0 due to license incompatibility

== 0.3.2 2010-09-24
* Added support for drawing circles, rectangles and ellipses

== 0.3.1 2010-09-15
* checkbox support added
* iText updated to 5.0.4
* documentation updated and typos fixed

== 0.3.0 2009-02-08

* Refactored the code.
* Added JRuby support
* Fixed some documentation

== 0.2.0 2007-11-08

* Use bytearray for output
* Added support for images

== 0.1.0 2007-07-13

* 1 major enhancement:
  * Initial release


spec/example.rb
require 'rubygems'
require 'pdf/stamper'


@pdf = PDF::Stamper.new(File.join(File.dirname(__FILE__), "test_template.pdf"))
@pdf.text :text_field01, "test"
@pdf.text :text_field02, "test2"
@pdf.image :button_field01, File.join(File.dirname(__FILE__), "logo.gif")

@pdf.save_as "test_output.pdf"

spec/pdf_stamper_spec.rb
$:.unshift File.join(File.dirname(__FILE__), "..", "lib")
require 'pdf/stamper'

describe PDF::Stamper do
  before(:each) do
    @pdf = PDF::Stamper.new(File.join(File.dirname(__FILE__), "test_template.pdf"))
    @pdf.text :text_field01, "test"
    @pdf.text :text_field02, "test2"
    @pdf.image :button_field01, File.join(File.dirname(__FILE__), "logo.gif")
  end

  it "should create PDF document" do
    @pdf.to_s.should_not be_nil
  end

  it "should save PDF document" do
    @pdf.save_as "test_output.pdf"
    File.exist?("test_output.pdf").should be_true
    File.delete("test_output.pdf") # Comment this out to view the output
  end
end


spec/pdf_stamper_rjb_spec.rb
$:.unshift File.join(File.dirname(__FILE__), "..", "lib")
require 'pdf/stamper/rjb'

describe PDF::Stamper do
  before(:each) do
    @pdf = PDF::Stamper.new(File.join(File.dirname(__FILE__), "test_template.pdf"))
    @pdf.text :text_field01, "test"
    @pdf.text :text_field02, "test2"
    @pdf.image :button_field01, File.join(File.dirname(__FILE__), "logo.gif")
  end

  it "should create PDF document" do
    @pdf.to_s.should_not be_nil
  end

  it "should save PDF document" do
    @pdf.save_as "test_output.pdf"
    File.exist?("test_output.pdf").should be_true
    File.delete("test_output.pdf") # Comment this out to view the output
  end
end


lib/pdf/stamper/rjb.rb
# = pdf/stamper/rjb.rb -- PDF template stamping.
#
#  Copyright (c) 2007-2012 Jason Yates

require 'rubygems'
require 'rjb'

Rjb::load(File.join(File.dirname(__FILE__), '..', '..', '..', 'ext', 'iText-4.2.0.jar'), ['-Djava.awt.headless=true'])

# PDF::Stamper::RJB
#
# RJB needs the LD_LIBRARY_PATH and JAVA_HOME environment set for it
# to work correctly.  For example on my system:
#
# export LD_LIBRARY_PATH=/usr/java/jdk1.6.0/jre/lib/i386/:/usr/java/jdk1.6.0/jre/lib/i386/client/:./
# export JAVA_HOME=/usr/java/jdk1.6.0/
#
# Check the RJB documentation if you are having issues with this.
module JRubyToRjb
  class << self; attr_reader :java_class; end
  def initialize(args)
    @object = self.java_class.new(args)
  end

  def method_missing(meth, *args, &block)
    @object.send meth, *args, &block
  end
end

class PdfReader
  include JRubyToRjb
  @java_class = Rjb::import('com.lowagie.text.pdf.PdfReader')
end

class PdfStamper
  include JRubyToRjb
  @java_class = Rjb::import('com.lowagie.text.pdf.PdfStamper')
end

class ByteArrayOutputSream
  include JRubyToRjb
  @java_class = Rjb::import('java.io.ByteArrayOutputStream')
end

class FileOutputStream
  include JRubyToRjb
  @java_class = Rjb::import('java.io.FileOutputStream')
end

class AcroFields
  include JRubyToRjb
  @java_class = Rjb::import('com.lowagie.text.pdf.AcroFields')
end

class PdfWriter
  include JRubyToRjb
  @java_class = Rjb::import('com.lowagie.text.pdf.PdfWriter')
end

class Image
  include JRubyToRjb
  @java_class = Rjb::import('com.lowagie.text.Image')
end


lib/pdf/stamper/jruby.rb
# = pdf/stamper/rjb.rb -- PDF template stamping.
#
#  Copyright (c) 2007-2009 Jason Yates

$:.unshift(File.join(File.dirname(__FILE__), '..', '..', '..', 'ext'))
require 'java'
require 'iText-4.2.0.jar'

java_import 'java.io.FileOutputStream'
java_import 'java.io.ByteArrayOutputStream'
java_import 'com.lowagie.text.pdf.AcroFields'
java_import 'com.lowagie.text.pdf.PdfReader'
java_import 'com.lowagie.text.pdf.PdfStamper'
java_import 'com.lowagie.text.Image'
java_import 'com.lowagie.text.Rectangle'
java_import 'com.lowagie.text.pdf.GrayColor'

module PDF
  include_package 'com.lowagie.text.pdf'

  class Stamper
    def initialize(pdf = nil)
      template(pdf) if ! pdf.nil?
    end
  
    def template(template)
      # NOTE I'd rather use a ByteArrayOutputStream.  However I
      # couldn't get it working.  Patches welcome.
      #@tmp_path = File.join(Dir::tmpdir, 'pdf-stamper-' + rand(10000).to_s + '.pdf')
      @reader = PDF::PdfReader.new(template)
      @baos = ByteArrayOutputStream.new
      @stamp = PDF::PdfStamper.new(@reader, @baos)#FileOutputStream.new(@tmp_path))
      @form = @stamp.getAcroFields()
      @black = GrayColor.new(0.0)
      @canvas = @stamp.getOverContent(1)
    end
    
    def is_checkbox(field_type)
      field_type == AcroFields::FIELD_TYPE_CHECKBOX
    end
  
    # Set a button field defined by key and replaces with an image.
    def image(key, image_path)
      # Idea from here http://itext.ugent.be/library/question.php?id=31 
      # Thanks Bruno for letting me know about it.
      img = Image.getInstance(image_path)
      img_field = @form.getFieldPositions(key.to_s)

      rect = Rectangle.new(img_field[1], img_field[2], img_field[3], img_field[4])
      img.scaleToFit(rect.width, rect.height)
      img.setAbsolutePosition(
        img_field[1] + (rect.width - img.scaledWidth) / 2,
        img_field[2] + (rect.height - img.scaledHeight) /2
      )

      cb = @stamp.getOverContent(img_field[0].to_i)
      cb.addImage(img)
    end

    def create_barcode(format)
      PDF.const_get("Barcode#{format}").new
    end

    def create_rectangle(coords)
      Rectangle.new(coords[1], coords[2], coords[3], coords[4])
    end

    def create_font(font_name)
      BaseFont.createFont(font_name, BaseFont.CP1252, false)
    end
    
    # Takes the PDF output and sends as a string.  Basically it's sole
    # purpose is to be used with send_data in rails.
    def to_s
      fill
      String.from_java_bytes(@baos.toByteArray)
    end
  end
end


lib/pdf/stamper.rb
# = pdf/stamper.rb -- PDF template stamping.
#
#  Copyright (c) 2007-2012 Jason Yates

require 'rbconfig'
require 'fileutils'
#require 'tmpdir'
#require 'active_support/inflector/methods'

include FileUtils

if RUBY_PLATFORM =~ /java/ # ifdef to check if your using JRuby
  require 'pdf/stamper/jruby'
else
  require 'pdf/stamper/rjb'
end

module PDF
  class Stamper
    VERSION = "0.6.0"
    
    
    # PDF::Stamper provides an interface into iText's PdfStamper allowing for the
    # editing of existing PDFs as templates. PDF::Stamper is not a PDF generator,
    # it allows you to edit existing PDFs and use them as templates.
    #
    # == Creation of templates
    #
    # Templates currently can be created using Adobe LiveCycle Designer
    # or Adobe Acrobat Professional. Using Acrobat Professional, you can create
    # a form and add textfields, checkboxes, radio buttons and buttons for images.
    #
    # == Example
    #
    # pdf = PDF::Stamper.new("my_template.pdf")
    # pdf.text :first_name, "Jason"
    # pdf.text :last_name, "Yates"
    # pdf.image :photo, "photo.jpg"
    # pdf.checkbox :hungry
    # pdf.save_as "my_output"

    def initialize(pdf = nil)
      template(pdf) if ! pdf.nil?
    end
  
    def template(template)
      reader = PdfReader.new(template)
      @baos = ByteArrayOutputStream.new
      @stamp = PdfStamper.new(reader, @baos)
      @form = @stamp.getAcroFields()
      @canvas = @stamp.getOverContent(1)
    end
  
    # Set a button field defined by key and replaces with an image.
    def image(key, image_path)
      # Idea from here http://itext.ugent.be/library/question.php?id=31 
      # Thanks Bruno for letting me know about it.
      img = Image.getInstance(image_path)
      img_field = @form.getFieldPositions(key.to_s)

      rect = Rectangle.new(img_field[1], img_field[2], img_field[3], img_field[4])
      img.scaleToFit(rect.width, rect.height)
      img.setAbsolutePosition(
        img_field[1] + (rect.width - img.scaledWidth) / 2,
        img_field[2] + (rect.height - img.scaledHeight) /2
      )

      cb = @stamp.getOverContent(img_field[0].to_i)
      cb.addImage(img)
    end
    
    # PDF::Stamper allows setting metadata on the created PDF by passing
    # the parameters to the set_more_info function. Our implementation here
    # is slightly different from iText, in that we only receive a single key/value
    # pair at a time, instead of a Map<string,string> since that is slightly
    # more complex to bridge properly from ruby to java.
    # 
    # Possible keys include "Creator". All values here are strings.
    # 
    def set_metadata(key, value)
      params = java.util.HashMap.new()
      params.put(key.to_s, value)
      @stamp.setMoreInfo(params)
    end
    
    # If you want to have iText reset some of the metadata, this function will
    # cause iText to use its default xml metadata.
    def reset_xmp_metadata()
      @stamp.setXmpMetadata("".to_java_bytes)
    end
    
    # Set a textfield defined by key and text to value
    def text(key, value)
      @form.setField(key.to_s, value.to_s) # Value must be a string or itext will error.
    end

    
    # Takes the PDF output and sends as a string.
    #
    # Here is how to use it in rails:
    #
    # def send 
    #     pdf = PDF::Stamper.new("sample.pdf") 
    #     pdf.text :first_name, "Jason"
    #     pdf.text :last_name, "Yates" 
    #     send_data(pdf.to_s, :filename => "output.pdf", :type => "application/pdf",:disposition => "inline")
    # end   
    def to_s
      fill
      String.from_java_bytes(@baos.toByteArray)
    end


    # Set a checkbox to checked
    def checkbox(key)
      field_type = @form.getFieldType(key.to_s)
      return unless is_checkbox(field_type)

      all_states = @form.getAppearanceStates(key.to_s)
      yes_state = all_states.reject{|x| x == "Off"}
      
      
      @form.setField(key.to_s, yes_state.first) unless (yes_state.size == 0)
    end
    
    # Get checkbox values
    def get_checkbox_values(key)
      field_type = @form.getFieldType(key.to_s)
      return unless is_checkbox(field_type)

      @form.getAppearanceStates(key.to_s)
    end

    def circle(x, y, r)
      @canvas.circle(x, y, r)
    end

    def ellipse(x, y, width, height)
      @canvas.ellipse(x, y, x + width, y + height)
    end

    def rectangle(x, y,  width, height)
      @canvas.rectangle(x, y, width, height)
    end

    # Example
    # barcode("PDF417", "2d_barcode", "Barcode data...", AspectRatio: 0.5)
    def barcode(format, key, value, opts = {})
      bar = create_barcode(format)
      bar.setText(value)
      opts.each do |name, opt|
        #bar.send("set#{name.to_s.camelize}", opt) #Camelize is not present outside of Rails by default
        bar.send("set#{name.to_s}", opt)
      end

      coords = @form.getFieldPositions(key.to_s)
      rect = create_rectangle(coords)

      barcode_img = bar.getImage
      barcode_img.scalePercent(100, 100 * bar.getYHeight)
      barcode_img.setAbsolutePosition(
          coords[1] + (rect.getWidth - barcode_img.getScaledWidth) / 2,
          coords[2] + (rect.getHeight - barcode_img.getScaledHeight) / 2
      )

      cb = @stamp.getOverContent(coords[0].to_i)
      cb.addImage(barcode_img)
    end

    # this has to be called *before* setting field values
    def set_font(font_name)
      itr = @form.getFields.keySet.iterator
      while itr.hasNext
        field = itr.next
        @form.setFieldProperty(field, 'textfont', create_font(font_name), nil)
      end
    end
    
    # Saves the PDF into a file defined by path given. If you want to save
    # to a string/buffer, just use .to_s directly.
    def save_as(file)
      File.open(file, "wb") { |f| f.write to_s }
    end
    
    private

    def fill
      @canvas.stroke()
      @stamp.setFormFlattening(true)
      @stamp.close
    end
  end
end
